metricbeat/module/linux/rapl/rapl.go

```diff
@@ -0,0 +1,261 @@
+	for _, cpu := range CPUList {
+		formatPath := filepath.Join(paths.Paths.Hostfs, "/dev/cpu/%d")
+		if config.UseMSRSafe {
+			formatPath = filepath.Join(formatPath, "/msr_safe")
+		} else {
+			formatPath = filepath.Join(formatPath, "/msr")
+		}
+		handler, err := rapl.CreateNewHandler(cpu, formatPath)
+		if err != nil {
+			return nil, errors.Wrapf(err, "error creating handler at path %s for CPU %d", formatPath, cpu)
+		}
+		handlers[cpu] = handler
+
+	}
+
+	ms := &MetricSet{
+		BaseMetricSet: base,
+		handlers:      handlers,
+	}
+
+	ms.updatePower()
+
+	return ms, nil
+}
+
+// Fetch methods implements the data gathering and data conversion to the right
+// format. It publishes the event which is then forwarded to the output. In case
+// of an error set the Error field of mb.Event or simply call report.Error().
+func (m *MetricSet) Fetch(report mb.ReporterV2) error {
+
+	watts := m.updatePower()
+
+	for cpu, metric := range watts {
+		evt := common.MapStr{
+			"core": cpu,
+		}
+		for domain, power := range metric {
+			evt[strings.ToLower(domain.Name)] = common.MapStr{
+				"watts":  common.Round(power.watts, common.DefaultDecimalPlacesCount),
+				"joules": common.Round(power.joules, common.DefaultDecimalPlacesCount),
+			}
+		}
+		report.Event(mb.Event{
+			MetricSetFields: evt,
+		})
+	}
+
+	return nil
+}
+
+func (m *MetricSet) updatePower() map[int]map[rapl.RAPLDomain]energyUsage {
+	newEnergy := make(map[int]map[rapl.RAPLDomain]energyTrack)
+	powerUsage := make(map[int]map[rapl.RAPLDomain]energyUsage)
+
+	for cpu, handler := range m.handlers {
+		powerUsage[cpu] = make(map[rapl.RAPLDomain]energyUsage)
+		domainList := map[rapl.RAPLDomain]energyTrack{}
+
+		for _, domain := range handler.GetDomains() {
+			joules, err := handler.ReadEnergyStatus(domain)
+			// This is a bit hard to check for, as many of the registers are model-specific
+			// Unless we want to maintain a map of every CPU, we sort of have to play it fast and loose.
+			if err == rapl.ErrMSRDoesNotExist {
+				continue
+			}
+			if err != nil {
+				logp.L().Infof("Error reading MSR from domain %s: %s skipping.", domain, err)
+				continue
+			}
+			domainList[domain] = energyTrack{joules: joules, time: time.Now()}
+			// divide the delta of joules by the time interval to get watts
+			if m.lastValues != nil {
+				// This register can roll over. If/when it does, skip reporting
+				if m.lastValues[cpu][domain].joules > joules {
+					continue
+				}
+				delta := m.lastValues[cpu][domain].joules - joules
+				timeDelta := m.lastValues[cpu][domain].time.Sub(domainList[domain].time)
+				powerUsage[cpu][domain] = energyUsage{watts: delta / timeDelta.Seconds(), joules: joules}
+			}
+		}
+		newEnergy[cpu] = domainList
+	}
+
+	m.lastValues = newEnergy
+	if m.lastValues == nil {
+		return nil
+	}
+
+	return powerUsage
+}
+
+// getMSRCPUs forms a list of CPU cores to query
+// For multi-processor systems, this will be more than 1.
+func getMSRCPUs() ([]int, error) {
```

Comment:
Oh, here's the function I mentioned in the other comment. So maybe you want to add here the
```if len(coreList) == 0 {
    return nil, errors.New("error getting list of cores. None found.")
}
```

Or something like that